File: implement_pi.py
import random
import numpy as np
    

epsilon = None # proba to not pick the best
alpha = None # learning rate

def update(SFs,now_sf,acked):
    global alpha
    SFs[now_sf-7] = SFs[now_sf-7] +alpha*(acked - SFs[now_sf-7])
    return SFs

def random_pick_sf(estimateval,minsf):
    global epsilon
    x = random.uniform(0,1)
    if (x > epsilon):
        return np.argmax(estimateval) + 7
    else :
        return random.choice([n for n in range (minsf,13)])


def main():
    
    is_acked = -1
    sf = 12
    minsf = 12

    SFs = [0]*6
    transmit(is_acked,sf,minsf,SFs)

def transmit(is_acked,sf,minsf,SFs):

    if is_acked > 0:
        SFs = update(SFs,sf,1)
    else:
        SFs = update(SFs,sf,0)
    sf = random_pick_sf(SFs,minsf)

File: implement_pi.c

//code à ajouter après les include de base {
#include <stdio.h>
#include <stdlib.h>
int epsilon = 0; // proba to not pick the best
int alpha = 1; // learning rate

int is_acked = -1;
int sf = 12;
int minsf = 12;

int SFs[6] = {0};

int max(int* tab){
    int indice = 0;
    int max = tab[0];
    for(int i = 0 ; i<sizeof(tab) ; i++){
        if(tab[i] > max){
            max = tab[i];
            indice = i;
        }
    }
    return indice;
}
//}


//code à ajouter avant les fonctions {
void update(int now_sf,int acked){
    SFs[now_sf-7] = SFs[now_sf-7] + alpha * (acked - SFs[now_sf-7]);
}

int random_pick_sf(int minsf){
    int x = rand() %2;
    if (x > epsilon)
        return max(SFs) + 7;
    else{
        int x = (rand() % (13-minsf)) + minsf;
        return x;
    }

}
//}




void transmit(){

    for(int i = 0; i<10; i++){

        //code à ajouter juste à la place du choix de sf {
        if (is_acked > 0)
            update(sf,1);
        else
            update(sf,0);
        sf = random_pick_sf(minsf); 
        is_acked = rand() % 2;
        //}
        

        printf("tentative %d : Sf = %d, acked = %d\n",i,sf,is_acked);
        for(int j =0; j<6;j++){
            printf("%d: [%d]\n",j,SFs[j]);
        }
    }
}

int main(){
    transmit();
    return 1;
}